### PR TITLE
Fix #1427 Add `DateTime.Should().Be(DateTime?)` and `DateTimeOffset.Should().Be(DateTimeOffset?)` overloads

### LINKED ISSUE #1427: Add `DateTime.Should().Be(DateTime?)` and `DateTimeOffset.Should().Be(DateTimeOffset?)` overloads
### Description

Some overloads of `Be` for primitive types have a nullable version to facilitate calling them against standard and nullable versions of the value. For example:

Normal:
image

Nullable:
image

However, these nullable overloads do not exist currently for both `DateTime` and `DateTimeOffset`.

I propose the overloads be added to help with testing these types of properties and for consistency in the FluentAssertions API.

### Expected behavior:

2 overloads be added, one for `DateTime` and another for `DateTimeOffset`, allowing to compare (`Be`) with their nullable counterparts. When the values are `null`, the assertion should obviously fail.

### Actual behavior:

Overloads do not exist, forcing extra logic inside assertions to get the `.Value`.

### Versions

- FluentAssertions 5.10.3
- Net Core 3.1

### PR DESCRIPTION
**Pull request by facundo91**
## Add `DateTime.Should().Be(DateTime?)` and `DateTimeOffset.Should().Be(DateTimeOffset?)` overloads

Fixes #1427 

To mimic the behavior of NumericAssertions, I have added the cases marked in bold to `DateTimeAssertions` and `DateTimeOffsetAssertions`. Below examples are for Datetime, but same logic was followed for DatetimeOffset.

### subject.Be(expected)

| subject       | expected      | result |
| -------       | --------      | ------ |
| 16/11/2020    | 16/11/2020    | true   |
|**16/11/2020**   | **null**          | **false** |
| null          | 16/11/2020    | false  |
| **null**          | **null**          | **true**   |

### subject.NotBe(expected)

| subject       | expected      | result |
| -------       | --------      | ------ |
| 16/11/2020    | 16/11/2020             | false  |
| **16/11/2020**    | **null**          | **true**   |
| null          | 16/11/2020    | true   |
| **null**          | **null**          | **false**  |

## IMPORTANT 

* [x] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [x] The changes are covered by a new or existing set of unit tests which follow the Arrange-Act-Assert syntax such as is used [in this example](https://github.com/fluentassertions/fluentassertions/blob/daaf35b9b59b622c96d0c034e8972a020b2bee55/Tests/FluentAssertions.Shared.Specs/BasicEquivalencySpecs.cs#L33).
* [x] If the contribution adds a feature or fixes a bug, please update [**the release notes**](https://github.com/fluentassertions/fluentassertions/tree/master/docs), which are published on the [website](https://fluentassertions.com/releases).
* [x] If the contribution changes the public API the changes needs to be included by running `AcceptApiChanges.ps1`/`AcceptApiChanges.sh`.
* [x] If the contribution affects [the documentation](https://github.com/fluentassertions/fluentassertions/tree/develop/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://www.fluentassertions.com/introduction).

### CODE REVIEW

File: Tests/FluentAssertions.Specs/Primitives/DateTimeAssertionSpecs.cs [hunk 1]
```diff
@@ -139,6 +139,20 @@ public void Should_succeed_when_asserting_datetime_value_is_equal_to_the_same_va
             act.Should().NotThrow();
         }
 
+        [Fact]
+        public void Should_succeed_when_asserting_datetime_value_is_equal_to_the_same_nullable_value()
+        {
+            // Arrange
+            DateTime dateTime = new DateTime(2016, 06, 04);
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Specs/Primitives/DateTimeAssertionSpecs.cs:146]: 🔧 Would be nice to rewrite this as `4.June(2016)`. Same for the other places.

File: Tests/FluentAssertions.Specs/Primitives/DateTimeAssertionSpecs.cs [hunk 2]
```diff
@@ -139,6 +139,20 @@ public void Should_succeed_when_asserting_datetime_value_is_equal_to_the_same_va
             act.Should().NotThrow();
         }
 
+        [Fact]
+        public void Should_succeed_when_asserting_datetime_value_is_equal_to_the_same_nullable_value()
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Specs/Primitives/DateTimeAssertionSpecs.cs:143]: 🙃 Although the code base is inconsistent, we prefer to use the `When_scenario_it_should_something`.

**Review by dennisdoomen — APPROVED** [review]
Nice PR. Just some nitpicking/suggestions if you're up to it.

File: Src/FluentAssertions/Primitives/DateTimeAssertions.cs [hunk 3]
```diff
@@ -65,6 +65,30 @@ public AndConstraint<TAssertions> Be(DateTime expected, string because = "", par
...
+        /// </param>
+        /// <param name="becauseArgs">
+        /// Zero or more objects to format using the placeholders in <paramref name="because" />.
+        /// </param>
+        public AndConstraint<TAssertions> Be(DateTime? expected, string because = "", params object[] becauseArgs)
+        {
+            Execute.Assertion
+                .ForCondition(
+                    (!Subject.HasValue && !expected.HasValue)
+                    || (Subject.HasValue && expected.HasValue && (Subject.Value == expected)))
```
[the remark this answers is not in the document]
Comment (facundo91) [reply Src/FluentAssertions/Primitives/DateTimeAssertions.cs]: Done!

File: Tests/FluentAssertions.Specs/Primitives/DateTimeAssertionSpecs.cs (hunk 1 above)
Comment (facundo91) [reply Tests/FluentAssertions.Specs/Primitives/DateTimeAssertionSpecs.cs]: Done!

File: Tests/FluentAssertions.Specs/Primitives/DateTimeAssertionSpecs.cs (hunk 2 above)
Comment (facundo91) [reply Tests/FluentAssertions.Specs/Primitives/DateTimeAssertionSpecs.cs]: Done! I think there should be a reference on the tests naming convention. Maybe in the Contributing guide.

**Comment by facundo91** [comment]
Thank you for the suggestions @dennisdoomen , it is now ready to be merge.

**Review by jnyrup — COMMENTED** [review]
Not crucial, but would be nice with tests exercising subject, expected or both being null.

File: Src/FluentAssertions/Primitives/DateTimeAssertions.cs [hunk 4]
```diff
@@ -65,6 +65,30 @@ public AndConstraint<TAssertions> Be(DateTime expected, string because = "", par
...
+        /// is needed. If the phrase does not start with the word <i>because</i>, it is prepended automatically.
+        /// </param>
+        /// <param name="becauseArgs">
+        /// Zero or more objects to format using the placeholders in <paramref name="because" />.
+        /// </param>
+        public AndConstraint<TAssertions> Be(DateTime? expected, string because = "", params object[] becauseArgs)
+        {
+            Execute.Assertion
+                .ForCondition(
+                    (!Subject.HasValue && !expected.HasValue) ||
```
[the remark this answers is not in the document]
Comment (facundo91) [reply Src/FluentAssertions/Primitives/DateTimeAssertions.cs:83]: Just tested, it fails at 
`FluentAssertions.Specs.DateTimeAssertionSpecs.Should_succeed_when_asserting_nullable_numeric_null_value_equals_null [FAIL]
      Did not expect any exception, but found System.InvalidOperationException with message "Nullable object must have a value."`

**Review by jnyrup — APPROVED** [review]

File: Src/FluentAssertions/Primitives/DateTimeAssertions.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jnyrup) [reply Src/FluentAssertions/Primitives/DateTimeAssertions.cs:83]: Hmm... I just pulled your branch and changed it to
```cs
.ForCondition(Subject == expected)
```

and all tests passed.

[the remark this answers is not in the document]
Comment (facundo91) [reply Src/FluentAssertions/Primitives/DateTimeAssertions.cs]: Yes, you are right. Just updated it.

**Pull request merged**